Task/short code adr (#4)
cds-snc/notification-adr: records/2021-09-20.scalibility.notify-short-code.md

# Add a short-code to Notify

Date: 2021-09-20

## Status

**IN REVIEW**

## Context

Short Codes are the foundation of mobile marketing. They allow consumers to interact with media outlets, brands, retailers, governments and all kinds of other organizations via text messaging. Instead of sending text messages to a 10-digit phone number, cell phone users send it to a Short Code – a five- or six-digit number that often spells a word or brand name. A Short Code becomes a “Common Short Code” (CSC) when the same code is activated across multiple mobile phone networks, which greatly extends its reach to many more mobile phone customers (this is from txt.ca). 

As a way to give Notify a brand, we looked into getting a short code for Notify. This would allow us to put the short code on Canada.ca, which would allow us to create trust in the service.

> Short codes allow customers to increase message throughput. Currently, shortcodes can send 100messages/ second in comparison to long codes that can send 1message/ second.

Reviewer: Let's provide a safe space to our wording.. 😅 

```suggestion
Short codes allow customers to increase message throughput. Currently, short codes can send 100 messages/ second in comparison to long codes that can send 1 message/ second.
```

1. Short codes are supposed to be tied to a single service. As can be seen from our application(and subsequent rejection) to AWS last year for short codes (https://docs.google.com/document/d/1IbBHE_hDtXmV03FJFeB5wRWUekzmEP4Ky2x0a_lDBtg/edit), each short code is meant to be single purpose.

> 2. Notify is a software service that is run by other services. The unsubscribe feature is managed by AWS (If a customer responds to the short code with 'STOP', AWS puts the user on a do not send list). AWS allows us to remove a single user from the unsubscribe list once every 30 days.

Reviewer: ```suggestion
2. Notify is a software service that is used by a multitude of governmental services. The unsubscribe feature is managed by AWS (If a customer responds to the short code with 'STOP', AWS puts the user on a do not send list). AWS allows us to remove a single user from the unsubscribe list once every 30 days.
```

Eg: Immigration, Refugee and Citizenship Canada (IRCC) and Canadian Border Services Agency (CBSA) use Notify for different usecases. The same short-code would be used by both services. This leads to problems with subscriptions, as an end user might chose to unsubscribe from notifications sent by IRCC but inadvertently blocks notifications sent by CBSA.
3. Short codes are very expensive. As of Sept 2021, a short code is 3000USD$ to setup and a 1000USD$ in recurring monthly costs. It would be cost prohibitive to get short codes per service that uses Notify.

## Potential Solutions

### Unsubscribe Option

The main problem that arises with short codes is that an individual might unsubscibe themselves from all notifications from the Canadian Government, vs the individual service they were interested in unsubscribing from.  

> We could send a shortened link to the user to try for unsubscribing vs a 'STOP' message. The unsubscribe list would be managed through the individual Notify services.

Reviewer: Let's note in the ADR that this is a problem we also have with long codes, it is not a new issue and is something we have to tackle even now. Hence this issue would not prevent us from adopting short code because we already have this issue unfortunately, if my understanding is correct.

#### Product work required:

> * Need to get a URL shortner that is Gov of Canada owned and used by multiple different departments.

Reviewer: Can we expand on why we'd need a URL shortener? Honest question here as I don't remember. 😅

> * Need to build out functionality for managing subscriptions within the Notify Admin.

Reviewer: We can mention the effort by platform for subscriptions management.

### Multiple Short Codes
Get multiple short codes - one per service. Currently, there are no Shared Short Codes in Canada. USA allows customers to use 10DLC (https://messagemedia.com/us/blog/shared-short-code-replaced-with-10dlc/) but there isn't an equivalent in Canada. Unless we are willing to use the same short code for every client, we should instead get a short code per client.

### Cost of Short Codes

> Short codes are expensive. Short codes require paperwork and 10-16 weeks to get processed. We do not have a self serve option for Notify clients to pick whether they want a Short Code or not. We would have to create a process that would inform individual services about the cost of obtaining a short code, as well as a process for application and applying the Short Code to our service.

Reviewer: 💸

### Increase Throughput
Currently we have multiple phone numbers to get over the 1sms/1second rule that is enforced by AWS.

> We do not have another option for increasing throughput unless we apply for the Short Code.

Reviewer: Ah we mention this here, but that should be in the context I think as this is a goal and not a solution per se.

## Decision

As this stage, without doing a more thorough review of the 'Unsubscribe Options', we are going to put the Short Code discussion on hold.

> This is mainly due to usability concerns associated with unsubscription and not having a product solution around managing them gracefully.

Reviewer: 👍

> Currently we have multiple phone numbers to get over the 1sms/1second rule that is enforced by AWS. We do not have another option for increasing throughput unless we apply for the Short Code. We would need to further dwell on a solution to this particular issue.

Reviewer: ```suggestion
Currently we have multiple phone numbers to get over the 1 sms / 1 second rule that is enforced by AWS. We do not have another option for increasing throughput unless we apply for the Short Code. We would need to further dwell on a solution to this particular issue.
```

### Unsubscribe Option

The main problem that arises with short codes is that an individual might unsubscibe themselves from all notifications from the Canadian Government, vs the individual service they were interested in unsubscribing from.

> This problem currently also exists with the long codes we use in Notify. The difference between the long code unsubscribe and the short code unsubscribe is that we have multiple long codes (user would have to unsubscribe from each of the long codes to be blocked by all GC services), vs a single short code that would block the user from all GC services.

Reviewer: I wish that was true but unfortunately, a recipient sending a `STOP` keyword to one of our SMS text will get blocked from any long code coming from the same AWS region (or country, unsure) and AWS account. 

That means that today, if a Notify recipient receives a SMS notification from a **non-dedicated** long code coming from `ca-central-1` region and replies `STOP`, we can't send new notifications coming from `ca-central-1` from our AWS Notify production account, even from a different number picked from the AWS pool of numbers.

On the other hand, we could still send that recipient a notification from a service with a **dedicated** long code, because we send these from the `us-west-2` region and the opt-out list associated with our AWS account is different. That is, there is one opt-out list per AWS region (or country, unsure) and AWS account (i.e. Notify prod). If a user replies with a `STOP` keyword again to one of our **dedicated** long code, then he'll get blocked for all **dedicated** long code.

So the opt-out list should not be impeding our adoption short codes because we already have the issue. This is great to describe the mechanism in this ADR but we need a few updates to be accurate.

We could send a shortened link to the user to try and get them to use the unsubscribe through the service vs 'STOP'. The unsubscribe list would be managed through the individual Notify services.

	- Need to get a URL shortner that is Gov of Canada owned and used by multiple different departments. The URL shortner would be used to send the unsubscribe link in the SMS so every service could manage its own subscriptions.

> 	- Need to build out functionality for managing subscriptions within the Notify Admin. There is currently work being done by another platform initiative regarding subscriptions, there might be an overlap of service functionality that could be used for notify subscription management.

Reviewer: 👍
